core/src/main/java/org/pentaho/di/metastore/MetaStoreConst.java

```diff
@@ -66,38 +70,128 @@ public class MetaStoreConst {
                 PluginServiceLoader.loadServices( MetastoreLocator.class );
-              metastoreLocator = metastoreLocators.stream().findFirst().orElse( null );
+              metastoreLocator = result = metastoreLocators.stream().findFirst().orElse( null );
             }
           } catch ( KettlePluginException e ) {
             logger.error( "Error getting metastore locator", e );
           }
         }
       }
-      return metastoreLocator;
+      return result;
+    }
+  };
+
+  private static final MetastoreDirectoryProvider NULL_METASTORE_DIRECTORY_PROVIDER =
+    new MetastoreDirectoryProvider() {
+      @Override
+      public IMetaStore getMetastoreForDirectory(String rootFolder) {
+        return null;
+      }
+    };
+
+  private static final Supplier<MetastoreDirectoryProvider> metastoreDirectoryProviderSupplier =
+    new Supplier<MetastoreDirectoryProvider> () {
+    private volatile MetastoreDirectoryProvider metastoreDirectoryProvider;
+
+    public MetastoreDirectoryProvider get() {
+      // double-checked idiom because this can fail until the plugin is loaded.
+      MetastoreDirectoryProvider result = metastoreDirectoryProvider;
+      if ( result == null ) {
+        synchronized ( this ) {
+          try {
+            if ( metastoreDirectoryProvider == null ) {
+              Collection<MetastoreDirectoryProvider> metastoreDirectoryProviders =
+                PluginServiceLoader.loadServices( MetastoreDirectoryProvider.class );
+              metastoreDirectoryProvider = metastoreDirectoryProviders.stream().findFirst().orElse( null );
+              if ( metastoreDirectoryProvider == null ) {
+                // cache the null so we don't constantly try to look this up.
+                metastoreDirectoryProvider = result = NULL_METASTORE_DIRECTORY_PROVIDER;
+              }
+            }
+          } catch ( KettlePluginException e ) {
+            logger.error( "Error getting metastore directory provider", e );
+          }
+        }
+      }
+      return result;
+    }
+  };
+
+  private static final MetastoreBowlProvider NULL_METASTORE_BOWL_PROVIDER =
+    new MetastoreBowlProvider() {
+      @Override
+      public IMetaStore getExplicitMetastoreForBowl(Bowl bowl) {
+        return null;
+      }
+
+      @Override
+      public IMetaStore getMetastoreForBowl(Bowl bowl) {
+        return null;
+      }
+    };
+
+  private static final Supplier<MetastoreBowlProvider> metastoreBowlProviderSupplier =
+    new Supplier<MetastoreBowlProvider> () {
+    private volatile MetastoreBowlProvider metastoreBowlProvider;
+
+    public MetastoreBowlProvider get() {
+      // double-checked idiom because this can fail until the plugin is loaded.
+      MetastoreBowlProvider result = metastoreBowlProvider;
+      if ( result == null ) {
+        synchronized ( this ) {
+          try {
+            if ( metastoreBowlProvider == null ) {
+              Collection<MetastoreBowlProvider> metastoreBowlProviders =
```

Comment:
Why not have a Bowl itself return a MetaStore? Instead of needing an additional provider that only works in tandem with a particular type of bowl
We could have the defaults in the default bowl, and the projects version in the plugin bowl without having to downcast bowls or search for additional services.

Reply:
Interesting idea. It would make the Bowl more of a programmatic context rather than just an identifier that gets passed around. I think it would not make the right separation of responsibilities. Not everything that needs a metastore should need a bowl. Not everything that implements a bowl should have a story for metastores. 

I'm similarly concerned about what will happen with the *second* use case for bowls. How will they overlap? Keeping bowls simpler will make it easier to refactor them for those additional use cases.

Reply:
Maybe we should have named it BowlId, I assumed a Bowl would contain things.
This wouldn't change whether something that needs a metastore needs a bowl. Getting the provider straight from the Bowl just removes the service lookup, and makes the use of bowls for metastore lookup explicit.
It can be optional, and not force any other Bowl to provide one.
Conversely, if a different bowl type wanted to have another metastore provider, it cannot do so with this implementation. There will only be one MetastoreBowlProvider, which recognizes only one type of Bowl. We'd need to start changing providers at bowl swaps for that to work.

Reply:
I mean, it has properties, like name and path, and whatever else we need it to have. 

I agree that the current API doesn't allow multiple implementations. this might be YAGNI, I'm not sure. 

I'm slightly concerned about consistency between the 3 different types of Project config: variables, metastore objects, and shared objects. Metastore is kinda the odd one out here, since the others primarily come through the AbstractMeta in most cases, but maybe BACKLOG-40276 will make it more similar for shared objects. 

Putting these objects on the Bowl would allow us to cache them better. I'm worried about repeated creations of XmlVfsMetastores, since we can't cache them effectively at the MetaStoreConst or lower level. 

If we did put these onto the Bowls, how would you expect stuff (like the UI) to access the global metastore? Explicitly using the DefaultBowl instance?

Reply:
I wouldn't expect much to change from this except for the service lookup to be replaced by using the bowl interface. Think it makes sense for uses to still go through their special manager/const classes to fit better with current logic and minimize impact.

Reply:
So basically, `Bowl` would have methods `getMetastore` and `getExplicitMetastore` and we'd implement those in `ProjectBowl` rather than having the `MetastoreBowlProvider` interface at all. And `MetaStoreConst` would just call `Bowl.getMetastore()` or `Bowl.getExplicitMetastore` instead of calling the `MetastoreBowlProvider`? 

I guess the weird part then would be, why would anyone call `MetaStoreConst.getMetastoreForBowlSupplier` if they already have a Bowl? (bearing in mind that there aren't many places that directly access the metastore anyway)

I think my biggest resistance to putting more stuff in the Bowl interface is based on wanting to keep as much "Project" stuff out of Kettle.  But also, I don't think it's a great idea to mix object identity and only vaguely-related concepts. If we go this way, should `Bowl` also implement `HasDatabases` and `VariableSpace` ? That seems wrong.

Reply:
If it wouldn't complicate things, I have nothing against getting things from the Bowl and knocking out a few singletons. But you understand the implications better than me.
For me, if something passes a bowl to something else to get a database, the bowl might as well be the one returning it. If that is going to be baked into kettle anyway, I'd rather look at a Bowl and have some idea of how it can affect things. Otherwise I'd have to sift through all uses to find what resources can depend on it.

Reply:
This change has been implemented.